src/pmu.rs

```diff
@@ -0,0 +1,292 @@
+    DataMisaligned,
+}
+
+#[derive(Debug)]
+pub enum CauseError {
+    InvalidCause,
+}
+
+impl PMUCfg {
+    ///
+    /// Resets SLEEP and WAKE programs on the PMU to defaults
+    ///
+    pub fn load_default_programs(&self) {
+        unsafe {
+            let pmu = PMU::ptr();
+
+            for i in 0..8 {
+                (*pmu).pmukey.write(|w| w.bits(PMU_KEY_VAL));
+                (*pmu).pmusleeppm[i].write(|w| w.bits(DEFAULT_SLEEP_PROGRAM[i]));
+
+                (*pmu).pmukey.write(|w| w.bits(PMU_KEY_VAL));
+                (*pmu).pmuwakepm[i].write(|w| w.bits(DEFAULT_WAKE_PROGRAM[i]));
+            }
+        }
+    }
+
+    ///
+    /// Puts device to sleep for N seconds, allowing wake-up button to wake it up as well
+    ///
+    /// # Arguments
+    /// 
+    /// *sleep_time* - the amount of time to sleep for in seconds
+    ///
+    /// # Notes
+    /// 
+    /// - enables RTC to be always on
+    /// - sets RTC scale to 1/s
+    ///
+    pub fn sleep(self, sleep_time: u32) {
+        unsafe {
+            let pmu = PMU::ptr();
+            let rtc = RTC::ptr();
+
+            // set interrupt source to RTC enabled, each pmu register needs key set before write
+            (*pmu).pmukey.write(|w| w.bits(PMU_KEY_VAL));
+            (*pmu)
+                .pmuie
+                .write(|w| w.rtc().set_bit().dwakeup().set_bit());
+            // set RTC clock scale to once per second for easy calculation
+            (*rtc)
+                .rtccfg
+                .write(|w| w.enalways().set_bit().scale().bits(15));
+            // get current RTC clock value scaled
+            let rtc_now = (*rtc).rtcs.read().bits();
+            // set RTC clock comparator
+            (*rtc).rtccmp.write(|w| w.bits(rtc_now + sleep_time));
+            // go to sleep for sleep_time seconds, need to set pmukey here as well
+            (*pmu).pmukey.write(|w| w.bits(PMU_KEY_VAL));
+            (*pmu).pmusleep.write(|w| w.sleep().set_bit());
+        }
+    }
+
+    ///
+    /// Stores user data `UD` to backup registers.
+    ///
+    /// # Arguments
+    ///
+    /// * `user_data` - the user data to store. *MUST* have alignment of at least 4 and fit into the backup registers
+    ///
+    /// # Returns
+    /// 
+    /// * `Result<UD, BackupError>` - the stored `user_data` is returned on success
+    /// 
+    /// # Errors
+    ///
+    /// * `BackupError::DataTooLarge` - returned if `user_data` cannot fit into backup registers
+    /// * `BackupError::DataMisaligned` - returned if `user_data` is not aligned to at least 4 bytes
+    ///
+    /// # Notes
+    ///
+    /// You can use `#[repr(align(4))]` to enforce a minimum alignment of 4 bytes for `user_data`
+    ///
+    pub fn store_backup<UD>(&self, user_data: UD) -> Result<UD, BackupError>
+    where
+        UD: Sized,
+    {
+        unsafe {
+            let backup = BACKUP::ptr();
+            let ud_size = core::mem::size_of::<UD>();
+
+            if ud_size > (*backup).backup.len() {
```

Comment:
As the size is static, might it be possible to assert this at compile time?

Reply:
As far as I know this is only possible with const generics which haven't yet landed in Rust stable yet. I also think that while sizeof will be check-able with something like `where Sized(s) => s < 40` or such, alignment will probably remain in the macro world making it hard to assert them both.

If you can find a way to do this in nightly atm. I'll gladly add a `TODO: revisit as compile-time checks once X lands in stable`

Reply:
no idea, unfortunately… had assumed that static assertions would be more of a thing in rust!

Reply:
another thing; `backup.len()` will be in words, while `ud_size` is still in bytes

Reply:
fixed, nice find!